File: src/test/java/io/protobase/f7/formulas/logic/EXACTTest.java
package io.protobase.f7.formulas.logic;

import io.protobase.f7.errors.DivException;
import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class EXACTTest extends TestFormula {

  @Test
  public void test_sameType() {
    assertThat(EXACT.SELF.apply(null, 10.0, 10.0)).isEqualTo(true);
    assertThat(EXACT.SELF.apply(null, 10.0, "10.0")).isEqualTo(false);
    assertThat(EXACT.SELF.apply(null, 10.0, "10")).isEqualTo(true);
    assertThat(EXACT.SELF.apply(null, 10.0, 0.0)).isEqualTo(false);
    assertThat(EXACT.SELF.apply(null, "Same", "Same")).isEqualTo(true);
    assertThat(EXACT.SELF.apply(null, "Same", "Different")).isEqualTo(false);
    assertThat(EXACT.SELF.apply(null, true, true)).isEqualTo(true);
    assertThat(EXACT.SELF.apply(null, true, false)).isEqualTo(false);
    assertThat(EXACT.SELF.apply(null, false, false)).isEqualTo(true);
  }

  @Test
  public void test_error() {
    assertThat(EXACT.SELF.apply(null, new ValueException(), new ValueException())).isEqualTo(new ValueException());
    assertThat(EXACT.SELF.apply(null, new DivException(), new ValueException())).isEqualTo(new DivException());
    assertThat(EXACT.SELF.apply(null, 4.4444, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void test_lookup() {
    EXACT F = new EXACT(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn("A");
    when(collateralLookup.apply(A1, G19_RANGE)).thenReturn("B");
    assertThat(F.apply(A1, M22_RANGE, G19_RANGE)).isEqualTo(false);
    verify(collateralLookup).apply(A1, M22_RANGE);
  }

  @Test
  public void test_grids() {
    Grid one = Grid.builder()
        .add(0, 0, 44.0)
        .add(0, 1, "A")
        .build();
    Grid two = Grid.builder()
        .add(0, 0, 44.0)
        .add(0, 1, "B")
        .build();
    assertThat(EXACT.SELF.apply(null, one, two)).isEqualTo(true);
  }

  @Test
  public void test_errorFromArgumentsMismatch() {
    assertThat(EXACT.SELF.apply(null, "Too few")).isEqualTo(new NAException());
    assertThat(EXACT.SELF.apply(null, "A", "B", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/logic/EXACTTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7Exception } from "../../../../main/js/errors/F7Exception";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { EXACT } from "../../../../main/js/formulas/logic/EXACT";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("EXACT", function () {
  it("should do same type equality", function () {
    assert.equal(EXACT.SELF.run(null, 10.0, 10.0), true);
    assert.equal(EXACT.SELF.run(null, 10.0, 0.0), false);
    assert.equal(EXACT.SELF.run(null, "Same", "Same"), true);
    assert.equal(EXACT.SELF.run(null, "Same", "Different"), false);
    assert.equal(EXACT.SELF.run(null, true, true), true);
    assert.equal(EXACT.SELF.run(null, true, false), false);
    assert.equal(EXACT.SELF.run(null, false, false), true);
  });

  it("should do pass-through errors", function () {
    assert.equal(
      (EXACT.SELF.run(null, 4.4444, new ValueException()) as F7Exception).name,
      F7ExceptionName.VALUE
    );
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new EXACT(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns("A");
    collateralLookup.withArgs(CommonModels.A1, CommonModels.G19_RANGE).returns("B");
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE, CommonModels.G19_RANGE), false);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.G19_RANGE));
    assert.equal(collateralLookup.callCount, 2);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 44.0).add(0, 1, "A").build();
    const two = Grid.builder().add(0, 0, 44.0).add(0, 1, "B").build();
    assert.deepEqual(EXACT.SELF.run(null, one, two), true);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((EXACT.SELF.run(null, "Too few") as NAException).name, F7ExceptionName.NA);
    assert.deepEqual(
      (EXACT.SELF.run(null, "A", "B", "Too many") as NAException).name,
      F7ExceptionName.NA
    );
  });
});
